File: jbake-core/src/test/java/org/jbake/app/ContentStoreIntegrationTest.java
package org.jbake.app;

import org.jbake.TestUtils;
import org.jbake.app.configuration.ConfigUtil;
import org.jbake.app.configuration.DefaultJBakeConfiguration;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.ClassRule;
import org.junit.rules.TemporaryFolder;

import java.io.File;

public abstract class ContentStoreIntegrationTest {

    @ClassRule
    public static TemporaryFolder folder = new TemporaryFolder();
    protected static ContentStore db;
    protected static DefaultJBakeConfiguration config;
    protected static File sourceFolder;

    @BeforeClass
    public static void setUpClass() throws Exception {

        sourceFolder = TestUtils.getTestResourcesAsSourceFolder();
        Assert.assertTrue("Cannot find sample data structure!", sourceFolder.exists());

        config = (DefaultJBakeConfiguration) new ConfigUtil().loadConfig(sourceFolder);
        config.setSourceFolder(sourceFolder);

        Assert.assertEquals(".html", config.getOutputExtension());
        db = new ContentStore();
    }

    @AfterClass
    public static void cleanUpClass() {
        // no-op
    }

    @Before
    public void setUp() {
        // no-op: in-memory store is always ready
    }

    @After
    public void tearDown() {
        db.drop();
    }
}

File: jbake-core/src/test/java/org/jbake/app/ContentStoreIntegrationTest.kt
package org.jbake.app

import io.kotest.matchers.shouldBe
import org.apache.commons.lang3.SystemUtils
import org.jbake.TestUtils
import org.jbake.app.configuration.ConfigUtil
import org.jbake.app.configuration.DefaultJBakeConfiguration
import org.jbake.app.DatabaseType
import java.io.File
import java.nio.file.Files

abstract class ContentStoreIntegrationTest {
    protected fun setUp() {
        db.startup()
    }

    protected fun tearDown() {
        db.drop()
    }

    companion object {
        internal lateinit var tempDir: File
        internal lateinit var db: ContentStore
        internal lateinit var config: DefaultJBakeConfiguration
        internal var databaseType: DatabaseType = DatabaseType.HSQLDB
        internal lateinit var sourceDir: File

        fun setUpClass(dbType: DatabaseType = DatabaseType.HSQLDB) {
            databaseType = dbType
            tempDir = Files.createTempDirectory("jbake-test-${dbType.storeName}").toFile()

            sourceDir = TestUtils.testResourcesAsSourceDir
            if (!sourceDir.exists()) throw AssertionError("Cannot find sample data structure!")

            config = ConfigUtil().loadConfig(sourceDir) as DefaultJBakeConfiguration
            config.setSourceDir(sourceDir)
            config.outputExtension shouldBe ".html"

            val dbPath = createDbPath(dbType)
            db = createContentStore(dbType, dbPath)
        }

        private fun createDbPath(dbType: DatabaseType): String {
            val timestamp = System.currentTimeMillis() - 1764000000000
            var dbPath = File(tempDir, "documents-${dbType.storeName}-$timestamp").name

            if (SystemUtils.IS_OS_WINDOWS)
                dbPath = dbPath.replace(":", "")

            return dbPath
        }

        private fun createContentStore(dbType: DatabaseType, dbPath: String): ContentStore = when (dbType) {
            DatabaseType.HSQLDB -> ContentStore("memory", dbPath)
            DatabaseType.NEO4J -> ContentStore("memory", dbPath)
            DatabaseType.ORIENTDB -> ContentStore("memory", dbPath)
        }

        fun cleanUpClass() {
            db.close()
            db.shutdown()
            tempDir.deleteRecursively()
        }
    }
}
